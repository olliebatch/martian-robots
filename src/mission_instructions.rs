use crate::robots::{Robot, RobotPosition, RobotStatus};
use anyhow::anyhow;
use std::collections::HashSet;
use std::fmt;
use std::str::FromStr;

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Coordinates {
    pub x: i32,
    pub y: i32,
}

impl Coordinates {
    pub fn fallen_off_grid(&self, coordinate_limit: &Coordinates) -> bool {
        if (self.x < 0 || self.x > coordinate_limit.x)
            || (self.y < 0 || self.y > coordinate_limit.y)
        {
            return true;
        }
        false
    }

    pub fn check_max_value(&self) -> Result<(), anyhow::Error> {
        if (self.x > 50 || self.x < 0) || (self.y > 50 || self.y < 0) {
            return Err(anyhow!(format!(
                "Grid provided is incorrect, {}, {}",
                self.x, self.y
            )));
        }
        Ok(())
    }
}

impl FromStr for Coordinates {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // remove any spaces if necessary
        let mut coords = s.split(' ');

        let x_from_char = coords.next().unwrap().to_string().parse::<i32>()?;
        let y_from_char = coords.next().unwrap().to_string().parse::<i32>()?;

        Ok(Coordinates {
            x: x_from_char,
            y: y_from_char,
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Command {
    pub upper_right: Coordinates,
    pub robots: Vec<Robot>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Orientation {
    North,
    South,
    West,
    East,
}

impl fmt::Display for Orientation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let val = match self {
            Orientation::North => "N",
            Orientation::South => "S",
            Orientation::West => "W",
            Orientation::East => "E",
        };
        write!(f, "{:}", val)
    }
}

impl FromStr for Orientation {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "N" => Ok(Orientation::North),
            "S" => Ok(Orientation::South),
            "W" => Ok(Orientation::West),
            "E" => Ok(Orientation::East),
            _ => Err(anyhow!("Error matching orientation")),
        }
    }
}
impl Orientation {
    pub fn change_right(&self) -> Self {
        match self {
            Orientation::North => Orientation::East,
            Orientation::East => Orientation::South,
            Orientation::South => Orientation::West,
            Orientation::West => Orientation::North,
        }
    }
    pub fn change_left(&self) -> Self {
        match self {
            Orientation::North => Orientation::West,
            Orientation::East => Orientation::North,
            Orientation::South => Orientation::East,
            Orientation::West => Orientation::South,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum RobotCommands {
    Left,
    Right,
    Forward,
}

impl RobotCommands {
    pub fn process(
        &self,
        robot_position: RobotPosition,
        coordinate_limit: &Coordinates,
        scent_tracker: &mut HashSet<RobotPosition>,
    ) -> (RobotPosition, RobotStatus) {
        match self {
            RobotCommands::Right => {
                let new_orientation = robot_position.orientation.change_right();
                (
                    robot_position.update_orientation(new_orientation),
                    RobotStatus::Alive,
                )
            }
            RobotCommands::Left => {
                let new_orientation = robot_position.orientation.change_left();
                (
                    robot_position.update_orientation(new_orientation),
                    RobotStatus::Alive,
                )
            }
            RobotCommands::Forward => {
                let scent_check = robot_position.check_scent(scent_tracker);
                if scent_check {
                    return (robot_position, RobotStatus::Alive);
                }
                let new_position = robot_position.move_forward();
                let fallen_off_grid = new_position.coordinates.fallen_off_grid(coordinate_limit);
                if fallen_off_grid {
                    robot_position.add_scent(scent_tracker);
                    return (robot_position, RobotStatus::Lost);
                }
                (new_position, RobotStatus::Alive)
            }
        }
    }
}

impl FromStr for RobotCommands {
    type Err = anyhow::Error;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        match input {
            "L" => Ok(RobotCommands::Left),
            "R" => Ok(RobotCommands::Right),
            "F" => Ok(RobotCommands::Forward),
            _ => Err(anyhow!("Error matching possible Robot Commands")),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
    use std::collections::HashSet;
    use std::str::FromStr;

    use crate::robots::Robot;
    use crate::RobotPosition;
    use rstest::*;

    #[rstest]
    #[case("1 2", 1, 2)]
    #[case("2 3", 2, 3)]
    #[case("11 21", 11, 21)]
    fn test_from_str_for_coords(
        #[case] input: &str,
        #[case] expected_x: i32,
        #[case] expected_y: i32,
    ) {
        let p = Coordinates::from_str(input);
        assert_eq!(
            p.unwrap(),
            Coordinates {
                x: expected_x,
                y: expected_y
            }
        )
    }

    #[rstest]
    #[case("N", Orientation::North)]
    #[case("W", Orientation::West)]
    #[case("E", Orientation::East)]
    #[case("S", Orientation::South)]
    fn test_orientation_from_str(#[case] input: &str, #[case] expected_orientation: Orientation) {
        let p = Orientation::from_str(input);
        assert_eq!(p.unwrap(), expected_orientation)
    }

    #[test]
    fn test_orientation_err_from_str() {
        let p = Orientation::from_str("Y");
        assert!(p.is_err())
    }

    #[rstest]
    #[case("F", RobotCommands::Forward)]
    #[case("L", RobotCommands::Left)]
    #[case("R", RobotCommands::Right)]
    fn test_robot_commands(#[case] input: &str, #[case] expected_command: RobotCommands) {
        let command = RobotCommands::from_str(input);

        assert_eq!(command.unwrap(), expected_command);
    }

    #[rstest]
    #[case(Orientation::North, Orientation::East)]
    #[case(Orientation::South, Orientation::West)]
    #[case(Orientation::East, Orientation::South)]
    #[case(Orientation::West, Orientation::North)]
    fn test_moving_right(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_right();
        assert_eq!(p, expected_orientation)
    }
    #[rstest]
    #[case(Orientation::North, Orientation::West)]
    #[case(Orientation::South, Orientation::East)]
    #[case(Orientation::East, Orientation::North)]
    #[case(Orientation::West, Orientation::South)]
    fn test_moving_left(#[case] input: Orientation, #[case] expected_orientation: Orientation) {
        let p = input.change_left();
        assert_eq!(p, expected_orientation)
    }

    #[rstest]
    #[case(RobotCommands::Forward)]
    fn test_processing_rob_commands_dont_move(#[case] input: RobotCommands) {
        let robot = Robot::new_basic_robot();

        let coordinate = Coordinates { x: 5, y: 3 };

        let mut scent_tracker: HashSet<RobotPosition> = HashSet::new();

        robot.position.add_scent(&mut scent_tracker);

        let result = input.process(robot.position, &coordinate, &mut scent_tracker);
        insta::assert_debug_snapshot!(result)
    }

    #[rstest]
    #[case(RobotCommands::Forward)]
    fn test_processing_rob_commands_fall_off_result(#[case] input: RobotCommands) {
        let robot = Robot::new_basic_robot();

        let coordinate = Coordinates { x: 1, y: 1 };

        let mut scent_tracker: HashSet<RobotPosition> = HashSet::new();

        let result = input.process(robot.position, &coordinate, &mut scent_tracker);
        insta::assert_debug_snapshot!(result)
    }

    #[rstest]
    #[case(RobotCommands::Forward)]
    fn test_processing_rob_commands_fall_off_scent(#[case] input: RobotCommands) {
        let robot = Robot::new_basic_robot();

        let coordinate = Coordinates { x: 1, y: 1 };

        let mut scent_tracker: HashSet<RobotPosition> = HashSet::new();

        let result = input.process(robot.position, &coordinate, &mut scent_tracker);
        insta::assert_debug_snapshot!(scent_tracker)
    }
}
