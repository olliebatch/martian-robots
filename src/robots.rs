use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub struct RobotPosition {
    pub coordinates: Coordinates,
    pub orientation: Orientation,
}

impl RobotPosition {
    pub fn update_orientation(self, orientation: Orientation) -> RobotPosition {
        RobotPosition {
            coordinates: self.coordinates,
            orientation,
        }
    }

    pub fn move_forward(&self) -> Self {
        let new_coordinates = match self.orientation {
            Orientation::North => Coordinates {
                x: self.coordinates.x,
                y: self.coordinates.y + 1,
            },
            Orientation::South => Coordinates {
                x: self.coordinates.x,
                y: self.coordinates.y - 1,
            },
            Orientation::East => Coordinates {
                x: self.coordinates.x + 1,
                y: self.coordinates.y,
            },
            Orientation::West => Coordinates {
                x: self.coordinates.x - 1,
                y: self.coordinates.y,
            },
        };
        RobotPosition {
            coordinates: new_coordinates,
            orientation: self.orientation.to_owned(),
        }
    }
}

impl fmt::Display for RobotPosition {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{:} {:} {:}",
            self.coordinates.x, self.coordinates.y, self.orientation
        )
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Robot {
    pub position: RobotPosition,
    pub robot_commands: Vec<RobotCommands>,
}

impl Robot {
    #[cfg(test)]
    pub fn new_basic_robot() -> Self {
        let robot_position = RobotPosition {
            coordinates: Coordinates { x: 1, y: 1 },
            orientation: Orientation::East,
        };
        let robot_commands = vec![
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
            RobotCommands::Right,
            RobotCommands::Forward,
        ];
        Robot {
            robot_commands,
            position: robot_position,
        }
    }

    pub fn new() -> Self {
        Robot {
            position: RobotPosition {
                coordinates: Coordinates { x: 0, y: 0 },
                orientation: Orientation::North,
            },
            robot_commands: vec![],
        }
    }

    pub fn set_start_position(self, robot_position: RobotPosition) -> Self {
        Robot {
            position: robot_position,
            robot_commands: self.robot_commands,
        }
    }

    pub fn update_commands(self, robot_commands: Vec<RobotCommands>) -> Self {
        Robot {
            position: self.position,
            robot_commands,
        }
    }

    pub fn process_robot_command(mut self, coordinate_limit: &Coordinates) -> Self {
        let command = self.robot_commands.first().unwrap();
        let new_position = command.process(self.position, coordinate_limit);
        self.robot_commands.remove(0);
        Robot {
            position: new_position,
            robot_commands: self.robot_commands,
        }
    }

    pub fn process_all_commands(mut self, coordinate_limit: &Coordinates) -> Self {
        while !self.robot_commands.is_empty() {
            self = self.process_robot_command(coordinate_limit)
        }
        self
    }
}

#[cfg(test)]
mod test {

    use crate::mission_instructions::{Coordinates, Orientation, RobotCommands};
    use crate::robots::{Robot, RobotPosition};
    use rstest::*;

    #[rstest]
    #[case(Orientation::North)]
    #[case(Orientation::South)]
    #[case(Orientation::West)]
    #[case(Orientation::East)]
    fn test_from_str_for_coords(#[case] orientation: Orientation) {
        let robot_position = RobotPosition {
            coordinates: Coordinates { x: 2, y: 2 },
            orientation,
        };

        let moved = robot_position.move_forward();
        insta::assert_debug_snapshot!(moved)
    }

    #[rstest]
    #[case(RobotCommands::Forward)]
    #[case(RobotCommands::Left)]
    #[case(RobotCommands::Right)]

    fn test_single_rob_command(#[case] command: RobotCommands) {
        let robot = Robot::new();
        let robot_with_commands = robot.update_commands(vec![command]);
        let processed_robot = robot_with_commands.process_robot_command();
        insta::assert_debug_snapshot!(processed_robot)
    }

    #[test]
    fn test_process_all_commands() {
        let new_basic_robot = Robot::new_basic_robot();
        let processed_robot = new_basic_robot.process_all_commands();
        insta::assert_debug_snapshot!(processed_robot)
    }
}
