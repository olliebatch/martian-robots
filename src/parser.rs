use crate::mission_instructions::{Command, Coordinates, Orientation, RobotCommands};
use crate::robots::{Robot, RobotPosition};
use anyhow::anyhow;
use itertools::Itertools;
use std::error;
use std::str::{FromStr, Lines};

pub fn parse_input_to_command(commands: &str) -> Result<Command, Box<dyn error::Error>> {
    println!("{:?}", commands);
    let mut lines_of_instruction = commands.lines();
    //todo add a check here to actually check the number of lines etc
    let coords = lines_of_instruction.next();
    // match until cover the lines so that can unwrap.
    let coordinates = match coords {
        None => Err(anyhow!("No Upper right point provided.")),
        Some(values) => Ok(Coordinates::from_str(values)?),
    }?;

    coordinates.check_max_value()?;

    let robots = parse_robot_commands(lines_of_instruction)?;

    let command = Command {
        upper_right: coordinates,
        robots,
    };

    Ok(command)
}

fn parse_robot_commands(lines: Lines) -> Result<Vec<Robot>, anyhow::Error> {
    let trimmed_lines = remove_lines_and_whitespace(lines);

    // assume current structure will stay the same with 2 lines = one robot
    let robot_infos = generate_robots_from_strs(trimmed_lines)?;

    Ok(robot_infos)
}

fn remove_lines_and_whitespace(lines: Lines) -> Vec<String> {
    //todo if time how to handle a robot that's dropped with no movements?
    let mut removed_lines: Vec<String> = vec![];
    for line in lines {
        if line.is_empty() {
            // do nothing because it's an empty line.
        } else {
            removed_lines.push(line.to_string())
        }
    }
    removed_lines
}

fn generate_robots_from_strs(trimmed_strings: Vec<String>) -> Result<Vec<Robot>, anyhow::Error> {
    let mut robots = vec![];
    for chunk in &trimmed_strings.into_iter().chunks(2) {
        let mut robot = Robot::new();
        for (index, robot_info) in chunk.into_iter().enumerate() {
            if index == 0 {
                let info = robot_info.rsplit_once(' ').unwrap();

                println!("{:?}", info);
                let coordinates = Coordinates::from_str(info.0)?;
                let orientation = Orientation::from_str(info.1)?;
                let robot_position = RobotPosition {
                    coordinates,
                    orientation,
                };
                robot = robot.set_start_position(robot_position);
            } else {
                let robot_commands: Result<Vec<RobotCommands>, anyhow::Error> = robot_info
                    .chars()
                    .map(|test| RobotCommands::from_str(test.to_string().as_str()))
                    .collect();
                robot = robot.update_commands(robot_commands?);
            }
        }
        robots.push(robot)
    }

    Ok(robots)
}

#[cfg(test)]
mod test {
    use crate::parser::{
        generate_robots_from_strs, parse_input_to_command, parse_robot_commands,
        remove_lines_and_whitespace,
    };

    #[test]
    fn test_parse_input_commands() {
        let str = "5 3\n3 2 N\nFRRFLLFFRRFLL";

        let robot_commands = parse_input_to_command(str).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_error_over_50() {
        let str = "51 51\n3 2 N\nFRRFLLFFRRFLL";

        let robot_commands = parse_input_to_command(str);

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parse_multiple_commands() {
        let str = "5 3\n1 1 E\nRFRFRFRF\n\n3 2 N\nFRRFLLFFRRFLL\n\n0 3 W\nLLFFFLFLFL";

        let robot_commands = parse_input_to_command(str).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parsing_robot_commands() {
        let str = "3 2 N\nFRRFLLFFRRFLL";
        let lines = str.lines();

        let robot_commands = parse_robot_commands(lines).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_parsing_multiple_robot_commands() {
        let str = "1 1 E\nRFRFRFRF\n\n3 2 N\nFRRFLLFFRRFLL";
        let lines = str.lines();

        let robot_commands = parse_robot_commands(lines).unwrap();

        insta::assert_debug_snapshot!(robot_commands)
    }

    #[test]
    fn test_remove_lines_and_whitespace_space_and_lines() {
        let str = "3 2 N\nFRRFLLFFRRFLL";
        let lines = str.lines();

        let removed_lines = remove_lines_and_whitespace(lines);
        let expected = vec!["3 2 N".to_string(), "FRRFLLFFRRFLL".to_string()];

        assert_eq!(removed_lines, expected)
    }

    #[test]
    fn test_remove_lines_and_whitespace_no_lines() {
        let str = "3 2 N\nFRRFLLFFRRFLL";
        let lines = str.lines();

        let removed_lines = remove_lines_and_whitespace(lines);
        let expected = vec!["3 2 N".to_string(), "FRRFLLFFRRFLL".to_string()];

        assert_eq!(removed_lines, expected)
    }

    #[test]
    fn test_generate_robots_from_strs() {
        let robots =
            generate_robots_from_strs(vec!["3 2 N".to_string(), "FRRFLLFFRRFLL".to_string()])
                .unwrap();

        insta::assert_debug_snapshot!(robots)
    }

    #[test]
    fn test_generate_robots_from_strs_two_robots() {
        let robots = generate_robots_from_strs(vec![
            "1 1 E".to_string(),
            "RFRFRFRF".to_string(),
            "3 2 N".to_string(),
            "FRRFLLFFRRFLL".to_string(),
        ])
        .unwrap();

        insta::assert_debug_snapshot!(robots)
    }
}
