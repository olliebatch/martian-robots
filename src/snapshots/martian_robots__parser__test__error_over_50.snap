---
source: src/parser.rs
assertion_line: 107
expression: robot_commands
---
Err(
    "Grid provided is incorrect, 51, 51",
)
